p2p/libp2p/interface.go

```diff
@@ -10,3 +14,17 @@ type ConnectionMonitor interface {
 	SetThresholdMinConnectedPeers(thresholdMinConnectedPeers int, netw network.Network)
 	ThresholdMinConnectedPeers() int
 }
+
+// Sharder defines the eviction computing process of unwanted peers
+type Sharder interface {
+	ComputeEvictList(pidList []peer.ID) []peer.ID
+	Has(pid peer.ID, list []peer.ID) bool
+	PeerShardResolver() p2p.PeerShardResolver
+	IsInterfaceNil() bool
+}
+
+// PeerDiscovererWithSharder extends the PeerDiscoverer with the possibility to set the sharder
+type PeerDiscovererWithSharder interface {
+	p2p.PeerDiscoverer
+	SetSharder(kadSharder interface{}) error
+}
```

Comment:
Still don't like interface{}

Reply:
There must be at least one similar function between the 2 implementations

Reply:
Added CommonInterface where it could not be avoided. Refactored some factories, removed TODOs